### ISSUE TITLE
Best practice - sharing data between components

### ISSUE THREAD

**Issue by maikeldaloo** [skatejs/skatejs#163]
Hi Trey,

I'm just curious if there's a recommended way to share data between components.
As an example, say I have the following 2 components:

`comments-container` and `comment`.

`comments-container` is responsible for loading comments from the server and rendering them with the class `comment` on the page.
That's where the `comment` component comes in and handles any comment-specific interactions.

At the moment, for the `comment` component I'm defining the `comments-container` element as a prototype attribute.
i.e.

```
skate('comment', {
  type: skate.type.CLASSNAME,
  prototype: {
    commentsContainer: document.getElementsByClassName('comments-container').item(0)
  }
  ...
});
```

I know that's one way of sharing data - but is it a method that you'd recommend?
If there's a better/nicer way, please share.

Thanks,
Maikel

**Comment by treshugart** [comment]
From what you've shown me, it seems like you want access to the container so that you can call methods on it. Either that or you want to get information from it on-demand from the comments themselves.

For the former, I would recommend emitting events from the children. When the event bubbles, the comment container can respond to it.

You can trigger native custom events by using the `Element.prototype.dispatchEvent()` and the `CustomEvent` constructor. Depending on your browser support requirements you'll have to polyfill the `CustomEvent` constructor and possibly the `dispatchEvent()` method (supported in IE9).

References:

- [dispatchEvent()](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent)
- [CustomEvent](https://developer.mozilla.org/en/docs/Web/API/CustomEvent)

For the latter situation, it depends on exactly what you want to do. Having to reach up to the parent generally means you need to decouple your code a bit more and possibly relocate responsibilities so that they can facilitate that decoupling.

One caveat to doing it the way you've described is that you've coupled how you're manipulating your data with your HTML structure. Parents control the children and if the children make assumptions about where they are, that will break if the parent decides that should change. The relationship here is cyclic and implicit.

Another caveat which exists because of the first is that functional programming dictates that information should flow in one direction. In the case of the DOM, parents should know about children but children should not know about their parents. This is why I had suggested bubbling events because that makes no assumption about parent heirarchy and the parent can respond to child events because they are supposed to know about children.

If you look at how the Shadow DOM style templating is designed, it allows the parent to choose where it's children should go, but never allow the child to decide where it should go on its own. How your components are designed should reflect this.

I hope this helps.

**Comment by maikeldaloo** [comment]
Yeah, makes sense.

I'm using [RadioJS](http://radio.uxder.com/documentation.html) for all my events needs, it's a pubsub library and I do use events whenever I need get the parent to perform an action, or any part of the app for that matter.

But if you were to access some properties on the `comments-container` component from your `comment` component, would you achieve it by querying the DOM and reading that attribute or is there's another recommended way?

My throught process is coming from AngularJS and how you can define a service between directives/controllers/etc which allow for easy data-sharing.

**Comment by treshugart** [comment]
Why not just use the built-in DOM events? The API has mostly stabilised across browsers since IE9 and you get bubbling for free. My biggest recommendation I can make to anyone using Skate or web components in general is: think native. It's been designed to work very well with existing APIs.

**Comment by maikeldaloo** [comment]
I actually used the CustomEvent class just this week for the first time, I was curious about it and had a quick peek.
For the current project, I'm under a bit of pressure, time-wise, and the RadioJS library was added in the beginning of the project.

I am, however, trying to use as much native as possible. My current project just happens to be the first project I build without jQuery or AngularJS. And I'm enjoying it a lot. So I'm heading into the more native direciton :)
